File: strings/validate_ip_address.py
class Solution:
    def validIPAddress(self, IP):
        if IP.count(".") == 3 and all(self.isIPv4(part) for part in IP.split(".")): return "IPv4"
        if IP.count(":") == 7 and all(self.isIPv6(part) for part in IP.split(":")): return "IPv6"
        return "Neither"

    def isIPv4(self, s):
        try:
            return str(int(s)) == s and 0 <= int(s) <= 255
        except ValueError:
            return False

    def isIPv6(self, s):
        if len(s) > 4: return False
        if s and s[0] == '-': return False

        try:
            dec = int(s, 16)
        except:
            dec = None

        return dec is not None and 0 <= dec <= 0xffff


class Solution:
    def validIPAddress(self, IP: str) -> str:
        grps = []
        idx = IP.find('.')
        ip_type = 'IPv4'

        if idx != -1:
            grps = IP.split('.')
            if len(grps) != 4: return 'Neither'
        else:
            grps = IP.split(':')
            if len(grps) != 8: return 'Neither'
            ip_type = 'IPv6'

        if ip_type == 'IPv4':
            for chunk in grps:
                if not chunk: return 'Neither'
                if not chunk.isnumeric(): return 'Neither'
                if len(chunk) > 1 and chunk[0] == '0': return 'Neither'
                if chunk and chunk[0] == '-': return 'Neither'

                num = int(chunk)
                if not 0 <= num <= 255: return 'Neither'
        else:
            for chunk in grps:
                if not chunk: return 'Neither'

                for c in chunk:
                    if not ('0' <= c <= '9' or 'a' <= c <= 'f' or 'A' <= c <= 'F'):
                        return 'Neither'

                if int(chunk, 16) < 0: return 'Neither'
                if len(chunk) > 4: return 'Neither'

        return ip_type

# Cleaner
class Solution:
    def validIPAddress(self, IP: str) -> str:
        grps = []
        idx = IP.find('.')
        ip_type = 'IPv4'

        if idx != -1:
            grps = IP.split('.')
            if len(grps) != 4: return 'Neither'
        else:
            grps = IP.split(':')
            if len(grps) != 8: return 'Neither'
            ip_type = 'IPv6'

        if ip_type == 'IPv4':
            if all(self.valid_ipv4_chunk(chunk) for chunk in grps): return 'IPv4'
        else:
            if all(self.valid_ipv6_chunk(chunk) for chunk in grps): return 'IPv6'

        return 'Neither'

    def valid_ipv4_chunk(self, chunk):
        if not chunk: return False
        if len(chunk) > 1 and chunk[0] == '0': return False
        if not chunk.isnumeric(): return False
        if chunk and chunk[0] == '-': return False

        num = int(chunk)
        if not 0 <= num <= 255: return False
        return True

    def valid_ipv6_chunk(self, chunk):
        if not chunk: return False

        for c in chunk:
            if not ('0' <= c <= '9' or 'a' <= c <= 'f' or 'A' <= c <= 'F'):
                return False

        if int(chunk, 16) < 0: return False
        if len(chunk) > 4: return False
        return True


# 468. Validate IP Address
# https://leetcode.com/problems/validate-ip-address/description/


# Intuition:
# Main things to check will be:
# 1. How many parts after split
# 2. Does each part adhere to the range in the decimal form or in string form itself to avoid type casting
# 3. 


sol = Solution()
assert sol.validIPAddress("2001:0db8:85a3:0:0:8A2E:0370:7334:") == 'Neither'




# Samples for defang
# assert sol.validIPAddress("127[.]0[.]0[.]1")
# assert sol.validIPAddress("127.0.0.1")
# assert sol.validIPAddress("127/./0[.]0[.]1")
# assert not sol.validIPAddress("127asdfasdf[.]0[.]0[.]256")

File: strings/validate_ip_address.rb
# # Best solution
# def valid_ip_address(ip)
#     if ip.match?(/\A\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\Z/)
#         ip_decimals = ip.split(".")
#         ip_decimals.each do |num|
#             return "Neither" if num.to_i > 255
#             return "Neither" if num.to_i == 0 && num.length > 1
#             return "Neither" if num.to_i != 0 && num[0] == '0'
#         end
#         return "IPv4"
#     elsif ip.match?(/\A([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}\Z/)
#         return "IPv6"
#     end
#     return "Neither"
# end


# def valid_ip_address_regex(ip)
#     return "IPv4" if /^(25[0-5]|2[0-4][0-9]|[1][0-9]{2}|[1-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[1][0-9]{2}|[1-9][0-9]?|[0-9])\.(25[0-5]|2[0-4][0-9]|[1][0-9]{2}|[1-9][0-9]?|[0-9])\.(25[0-5]|2[0-4][0-9]|[1][0-9]{2}|[1-9][0-9]?|[0-9])$/ === ip
#     return "IPv6" if /^[a-fA-F0-9]{1,4}\:[a-fA-F0-9]{1,4}\:[a-fA-F0-9]{1,4}\:[a-fA-F0-9]{1,4}\:[a-fA-F0-9]{1,4}\:[a-fA-F0-9]{1,4}\:[a-fA-F0-9]{1,4}\:[a-fA-F0-9]{1,4}$/ === ip
#     return "Neither"
# end


def valid_ip_address_detailed(ip)
    tryIP4 = ip.split(".")
    return 'IPv4' if tryIP4.length == 4 && ip[0] != "." && ip[-1] != "." && validateIPv4(tryIP4)

    tryIP6 = ip.split(":")
    return 'IPv6' if tryIP6.length == 8 && ip[0] != ":" && ip[-1] != ":" && validateIPv6(tryIP6)

    "Neither"
end

def validateIPv4(groups)
    groups.each do |group|
        return false if group[0] == '0' && group != '0'
        return false if group.length < 1
        return false if group.to_i > 255
        digits = '1234567890'.chars
        group.each_char {|char| return false if !digits.include?(char)}
    end
    true
end

def validateIPv6(groups)
    groups.each do |group|
        return false if group.empty? || group.length > 4
        chars = "1234567890abcdefABCDEF".chars
        group.each_char {|char| return false if !chars.include?(char)}
    end
    true
end

# def validate_ip(ip)
#     parts = ip.split('.')
#     return false if parts.size != 4
#     return parts.map { |part| valid_part?(part) }.all?
# end

# def valid_part?(part)
#     ip = ''
#     part.each_char do |c|
#         ip += c if c.between?('0', '9')
#     end

#     ip.to_i.between?(0, 255)
# end




# 468. Validate IP Address
# https://leetcode.com/problems/validate-ip-address/description/


require 'test/unit'
extend Test::Unit::Assertions

# assert_equal(valid_ip_address("2001:0db8:85a3:0:0:8A2E:0370:7334:"), 'Neither')

assert_equal(validate_ip("127[.]0[.]0[.]1"), true)
assert_equal(validate_ip("127.0.0.1"), true)
assert_equal(validate_ip("127/./0[.]0[.]1"), true)
assert_equal(validate_ip("127asdfasdf[.]0[.]0[.]256"), false)
